From 2ea1f6bd3dc5c0aca4a4ef550e1fa92b72c9f7f0 Mon Sep 17 00:00:00 2001
From: "Dr. Greg Wettstein" <greg@enjellic.com>
Date: Sat, 14 Oct 2017 10:35:45 -0500
Subject: [PATCH 30/37] Add global process id to event information.

In order to support the ability of the userspace modeling engine
to 'jail' a process which has exibited an extra-dimensional
behavior the global pid of the process generating the information
exchange event is now prependent to the ASCII event definition.

The PID is encoded in the following format:

pid{NNN}

Where NNN is the global process id.  The global pid is used
rather then the namespace version of the id in order to allow the
supervising daemon to issue a jailing request to the process from
a global context.
---
 security/integrity/ima/ima_identity.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/security/integrity/ima/ima_identity.c b/security/integrity/ima/ima_identity.c
index cc65432d417d..ef585cb0faf7 100644
--- a/security/integrity/ima/ima_identity.c
+++ b/security/integrity/ima/ima_identity.c
@@ -91,6 +91,7 @@ struct trajectory {
 	struct list_head list;
 	char *filename;
 	char *process;
+	pid_t pid;
 	struct actor_identity actor;
 	struct subject_identity subject;
 };
@@ -329,6 +330,7 @@ static struct trajectory * create_point(const char *filename,
 
 	strcpy(entry->filename, filename);
 	strcpy(entry->process, current->comm);
+	entry->pid = task_pid_nr(current);
 	entry->actor = *actor;
 	entry->subject = *subject;
 
@@ -691,7 +693,8 @@ static ssize_t show_ns_update(struct kobject *kobj,
 				       mp->u.measurement);
 			break;
 		case exchange_event:
-			retn = sprintf(page, "exchange event{%s:%s} actor{uid=%d, euid=%d, suid=%d, gid=%d, egid=%d, sgid=%d, fsuid=%d, fsgid=%d, cap=0x%llx} subject{uid=%d, gid=%d, mode=0%o, name_length=%u, name=%*phN, s_id=%s, s_uuid=%*phN, digest=%*phN}\n",
+			retn = sprintf(page, "exchange pid{%u} event{%s:%s} actor{uid=%d, euid=%d, suid=%d, gid=%d, egid=%d, sgid=%d, fsuid=%d, fsgid=%d, cap=0x%llx} subject{uid=%d, gid=%d, mode=0%o, name_length=%u, name=%*phN, s_id=%s, s_uuid=%*phN, digest=%*phN}\n",
+				       mp->u.exchange->pid,
 				       mp->u.exchange->process,
 				       mp->u.exchange->filename,
 				       mp->u.exchange->actor.uid,
-- 
2.16.2

